src/main/scala/com/microsoft/hyperspace/index/rules/FilterIndexRule.scala

```diff
@@ -53,19 +54,24 @@ object FilterIndexRule
       case ExtractFilterNode(originalPlan, filter, outputColumns, filterColumns) =>
         try {
           val candidateIndexes =
-            findCoveringIndexes(filter, outputColumns, filterColumns)
+            findCoveringIndexes(filter, outputColumns, filterColumns, plan)
           FilterIndexRanker.rank(spark, filter, candidateIndexes) match {
             case Some(index) =>
-              // As FilterIndexRule is not intended to support bucketed scan, we set
-              // useBucketUnionForAppended as false. If it's true, Hybrid Scan can cause
-              // unnecessary shuffle for appended data to apply BucketUnion for merging data.
-              val transformedPlan =
-                RuleUtils.transformPlanToUseIndex(
-                  spark,
-                  index,
-                  originalPlan,
-                  useBucketSpec = HyperspaceConf.useBucketSpecForFilterRule(spark),
-                  useBucketUnionForAppended = false)
+              // Detect whether the index contains nested fields.
+              val indexHasNestedColumns = (index.indexedColumns ++ index.includedColumns)
+                .exists(ResolverUtils.ResolvedColumn(_).isNested)
+              val ruleHelper = if (indexHasNestedColumns) {
```

Comment:
instead of this, how about defining 

```scala
object IndexRelationApplyHelper {
  def apply(index: IndexLogEntry): IndexRelationApplyHelper
}
```